llvm/lib/Target/AIE/AIEWawRegRewriter.cpp

```diff
@@ -0,0 +1,386 @@
+class AIEWawRegRewriter : public MachineFunctionPass {
+
+public:
+  static char ID;
+  AIEWawRegRewriter() : MachineFunctionPass(ID) {}
+
+  void getAnalysisUsage(AnalysisUsage &AU) const override {
+    AU.setPreservesCFG();
+    AU.addRequired<VirtRegMap>();
+    AU.addPreserved<VirtRegMap>();
+    // no Machine Instructions are added, therefore the SlotIndexes remain
+    // constant and preserved
+    AU.addRequired<SlotIndexes>();
+    AU.addPreserved<SlotIndexes>();
+    // no new Virtual Registers are generated, therefore the LiveDebugVariables
+    // do not have to be updated
+    AU.addRequired<LiveDebugVariables>();
+    AU.addPreserved<LiveDebugVariables>();
+    AU.addRequired<LiveStacks>();
+    AU.addPreserved<LiveStacks>();
+    AU.addRequired<LiveIntervals>();
+    AU.addPreserved<LiveIntervals>();
+    AU.addRequired<LiveRegMatrix>();
+    AU.addPreserved<LiveRegMatrix>();
+    MachineFunctionPass::getAnalysisUsage(AU);
+  }
+
+  bool runOnMachineFunction(MachineFunction &Fn) override;
+
+private:
+  MachineFunction *MF = nullptr;
+  const MachineRegisterInfo *MRI = nullptr;
+  const AIEBaseRegisterInfo *TRI = nullptr;
+  VirtRegMap *VRM = nullptr;
+  LiveRegMatrix *LRM = nullptr;
+  LiveIntervals *LIS = nullptr;
+
+  bool renameMBBPhysRegs(const MachineBasicBlock *MBBs);
+
+  /// Get all the defined physical registers that the MachineBasicBlock already
+  /// uses. These physical registers should not be used for replacement
+  /// candidates, since this would introduce new WAW dependencies, which this
+  /// pass tries to remove.
+  BitVector getDefinedPhysRegs(const MachineBasicBlock *MBB) const;
+
+  bool replaceReg(const Register Reg, BitVector &BlockedPhysRegs);
+
+  /// Find a free register of the same register class type, but
+  /// exclude the blocked physical registers from the result.
+  /// Otherwise a new WAW dependencies can be introduced, that was previously
+  /// removed.
+  MCPhysReg getReplacementPhysReg(const Register Reg,
+                                  const BitVector &BlockedPhysRegs) const;
+
+  bool isWorthRenaming(const Register &Reg, const BitVector &UsedPhysRegs,
+                       const SmallVector<Register, 16> &VRegWithCopies) const;
+
+  /// return the Physical register of the machine operand.
+  MCPhysReg getAssignedPhysReg(const MachineOperand &MO) const;
+
+  bool isIdentityCopy(const MachineInstr &MI) const;
+
+  /// MachineInstructions can write into sub-registers (%0:sub_256_lo and
+  /// %0:sub_256_hi). In the VirtualRegisterMap, they block the same
+  /// register (i.e. $x0). This can cause issues, when detecting Write
+  /// After Write (WAW) dependencies. To mitigate this, only use the last
+  /// definition of a virtual register to count the definitions, so that
+  /// writing into subregisters does not already trigger the register
+  /// replacement. In this case $x0 would have already been replaced, even
+  /// though there is no real WAW dependency.
```

Comment:
I don't think it was said before and if it was I missed it, but that may be worth calling out here or somewhere else that the reason you are doing this is because you detect that there is a WAW when there are more than one def and sub-regs can create more than one (partial) def which are by construction not WAW.

What I'm saying is your comment makes sense when the reader has this context in their head and without it, it may be hard to understand why we do that.

Reply:
Looking through more of the PR, I don't actually see that you don't do anything special with the subregs.
I see that we only do the rewriting when we reach the last definition of a virtual reg that has been defined at least twice, but both definitions can be partial (i.e., subreg) ones.

Am I missing something?
Put differently, if I'm right, this comment is confusing since it doesn't reflect the reality.

Reply:
Talked with @F-Stuckmann offline, I missed the use of virtual registers and physical registers in the checks that happens after the call to getLastVRegDef.